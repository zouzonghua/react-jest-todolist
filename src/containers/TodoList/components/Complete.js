import React, { Component } from 'react';

class Complete extends Component {
  render() {
    const { list, deleteCompleteItem,undoCompleteItem } = this.props;
    return (
      <div className="complete-list">
        <div className="complete-list-title">
          已完成
          <div className="complete-list-count" data-test="count">
            {list && list.length}
          </div>
        </div>
        <ul className="complete-list-content">
          {list &&
            list.map((item, index) => (
              <li className="complete-list-item" data-test="list-item" key={index}>
                <div>
                  <input
                    className="complete-list-item-checkbox"
                    checked={false}
                    onChange={(e) => {
                      e && e.stopPropagation();
                      e.target.checked && undoCompleteItem(index)
                    }}
                    type="checkbox"
                    data-test="checkbox"
                  />
                  <span>{item.value}</span>
                </div>
                <span
                  className="complete-list-delete"
                  onClick={(e) => {
                    e && e.stopPropagation();
                    deleteCompleteItem(index);
                  }}
                  data-test="delete-item"
                >
                  -
                </span>
              </li>
            ))}
        </ul>
      </div>
    );
  }
}

export default Complete;
