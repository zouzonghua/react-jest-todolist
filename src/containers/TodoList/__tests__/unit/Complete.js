import React from 'react';
import { shallow } from 'enzyme';
import Complete from '../../components/Complete';
import { findTestWrapper } from '../../../../utils/testUtils'

describe('Complete 组件测试', () => {

  it('样式正常', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<Complete list={listData} />)
    expect(wrapper).toMatchSnapshot()
  });

  it('完成列表当数据为空数组 count 数目为0，列表无内容 ', () => {
    const wrapper = shallow(<Complete list={[]} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItems = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('0');
    expect(listItems.length).toEqual(0);
  });


  it('完成列表当数据为空数组 count 数目显示数据长度，列表不为空 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<Complete list={listData} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItems = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('2');
    expect(listItems.length).toEqual(2);
  });

  it('完成列表当数据有内容时，要存在删除按钮 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<Complete list={listData} />);
    const deleteItems = findTestWrapper(wrapper, 'delete-item');
    expect(deleteItems.length).toEqual(2);
  });


  it('完成列表当数据有内容时，点击某个删除按钮，会调用 deleteCompleteItem 方法 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<Complete deleteCompleteItem={fn} list={listData} />);
    const deleteCompleteItems = findTestWrapper(wrapper, 'delete-item');
    deleteCompleteItems.at(index).simulate('click');
    expect(fn).toHaveBeenLastCalledWith(index);
  });


  it('完成列表当数据有内容时，要存在复选框按钮 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const wrapper = shallow(<Complete list={listData} />);
    const checkboxItems = findTestWrapper(wrapper, 'checkbox');
    expect(checkboxItems.length).toEqual(2);
  });


  it('完成列表当数据有内容时，点击某个复选框按钮，会调用 undoCompleteItem 方法 ', () => {
    const listData = [
      { status: 'div', value: '学习 jest' },
      { status: 'div', value: '学习 TDD' },
    ];
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<Complete undoCompleteItem={fn} list={listData} />);
    const checkboxCompleteItems = findTestWrapper(wrapper, 'checkbox');
    checkboxCompleteItems.at(index).simulate('change', {target: {checked: true}, stopPropagation: jest.fn()});
    expect(fn).toHaveBeenLastCalledWith(index);
  });
})
