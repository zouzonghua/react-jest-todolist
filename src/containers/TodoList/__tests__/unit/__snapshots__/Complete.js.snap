// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complete 组件测试 样式正常 1`] = `
<div
  className="complete-list"
>
  <div
    className="complete-list-title"
  >
    已完成
    <div
      className="complete-list-count"
      data-test="count"
    >
      2
    </div>
  </div>
  <ul
    className="complete-list-content"
  >
    <li
      className="complete-list-item"
      data-test="list-item"
      key="0"
    >
      <div>
        <input
          checked={false}
          className="complete-list-item-checkbox"
          data-test="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        <span>
          学习 jest
        </span>
      </div>
      <span
        className="complete-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
        -
      </span>
    </li>
    <li
      className="complete-list-item"
      data-test="list-item"
      key="1"
    >
      <div>
        <input
          checked={false}
          className="complete-list-item-checkbox"
          data-test="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        <span>
          学习 TDD
        </span>
      </div>
      <span
        className="complete-list-delete"
        data-test="delete-item"
        onClick={[Function]}
      >
        -
      </span>
    </li>
  </ul>
</div>
`;
